_posts/2022-09-15

```diff
@@ -0,0 +1,128 @@
+
+1. 1
+1. 1
+1. 1
+
+Similar to the Microsoft implementation, LightInject is also able to detect resolved instances implementing `IDisposable`. But LightInject avoids disposing the same instance multiple times.
+
+We can confirm that (unsurprisingly) both containers follow the [dependency injection guidelines](https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines) which states:
+
+> The container is responsible for cleanup of types it creates, and calls Dispose on IDisposable instances. Services resolved from the container should never be disposed by the developer.
+
+## A word of warning
+
+When working with _scoped_ lifetime, we can rely on the DI container disposing the instances created. The previously mentioned guidelines also apply to services resolved in _transient_ scope though. When working with dispsosable services in _transient_ scope, carefully consult your DI containers documentation as the behavior there is much less consistent across containers, e.g. the [Ninject documentation](https://github.com/ninject/Ninject/wiki/Object-Scopes) states that transient services won't be disposed by the container. The Microsoft guidelines also explicitly recommend to not register `IDisposable` services in the transient lifetime:
+
+> Don't register IDisposable instances with a transient lifetime. Use the factory pattern instead.
+
+## Handling the ambuigity
+
+Although we've seen that different containers might behave a bit different when disposing services, this should remain a fairly irrelevant detail in most cases. The Microsoft [guidelines on implementing `Dispose`](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose) states:
+
+> To help ensure that resources are always cleaned up appropriately, a Dispose method should be idempotent, such that it is callable multiple times without throwing an exception. Furthermore, subsequent invocations of Dispose should do nothing.
+
+As long as we stick to this guidline, the different implementations shouldn't affect the correctness of your system.
+
+
+## Conclusion
+
+Dependency injection containers do not only automatically dispose dependencies that directly implement `IDisposable` but will also take care of any disposable service implementation that is not visible to the consumer of the dependency. The exact behavior of the container in regards to disposing resolved instances can vary. For more complicated dependency chains, it can be tricky to predict the exact amount of times `Dispose` will be called on your class. However, this should ultimately belong in the "useless knowledge" category, because if you follow the general .NET coding guidelines, `Dispose` should be idempotent. That said, rest assured that there are plenty of implementations out there that might run into isuses in the shown examples (there is a reason this blog post exists ;)).
```

Comment:
```suggestion
Dependency injection containers do not only automatically dispose dependencies directly implementing the `IDisposable` interface. They will also take care of any disposable service implementation that is not visible to the consumer of the dependency.

The exact behavior of the container in regards to disposing resolved instances can vary. For more complicated dependency chains, predicting the number of times `Dispose` will be called can be tricky.

However, this should ultimately belong in the "useless knowledge" category because if you follow the general .NET coding guidelines, `Dispose` should be idempotent. That said, rest assured that plenty of implementations out there might run into issues in the shown examples (there is a reason this blog post exists ;)).
```